sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastRangeJoin.scala

```diff
@@ -0,0 +1,411 @@
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.joins
+
+import org.apache.spark.sql.catalyst.util.TypeUtils
+import org.apache.spark.util.ThreadUtils
+
+import scala.annotation.tailrec
+import scala.collection.mutable
+import scala.concurrent._
+import scala.concurrent.duration._
+
+import org.apache.spark.annotation.DeveloperApi
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.Row
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions._
+import org.apache.spark.sql.execution.{BinaryNode, SparkPlan}
+
+/**
+ * Performs an inner range join on two tables. A range join typically has the following form:
+ *
+ * SELECT A.*
+ *        ,B.*
+ * FROM   tableA A
+ *        JOIN tableB B
+ *         ON A.start <= B.end
+ *          AND A.end > B.start
+ *
+ * The implementation builds a range index from the smaller build side, broadcasts this index
+ * to all executors. The streaming side is then matched against the index. This reduces the number
+ * of comparisons made by log(n) (n is the number of records in the build table) over the
+ * typical solution (Nested Loop Join).
+ *
+ * TODO NaN values
+ * TODO NULL values
+ * TODO Outer joins? StreamSide is quite easy/BuildSide requires bookkeeping and
+ * TODO This join will maintain sort order. The build side rows will also be added in a lower
+ *      bound sorted fashion.
+ */
+@DeveloperApi
+case class BroadcastRangeJoin(
+    leftKeys: Seq[Expression],
+    rightKeys: Seq[Expression],
+    equality: Seq[Boolean],
+    buildSide: BuildSide,
+    left: SparkPlan,
+    right: SparkPlan)
+  extends BinaryNode {
+
+  private[this] lazy val (buildPlan, streamedPlan) = buildSide match {
+    case BuildLeft => (left, right)
+    case BuildRight => (right, left)
+  }
+
+  private[this] lazy val (buildKeys, streamedKeys) = buildSide match {
+    case BuildLeft => (leftKeys, rightKeys)
+    case BuildRight => (rightKeys, leftKeys)
+  }
+
+  override def output: Seq[Attribute] = left.output ++ right.output
+
+  @transient
+  private[this] lazy val buildSideKeyGenerator: Projection =
+    newProjection(buildKeys, buildPlan.output)
+
+  @transient
+  private[this] lazy val streamSideKeyGenerator: () => MutableProjection =
+    newMutableProjection(streamedKeys, streamedPlan.output)
+
+  private[this] val timeout: Duration = {
```

Comment:
maybe we should just put this logic in `SQLConf` so we can use it in both places.